File: src/detection/detect.ts
import { detectBrowserify } from "./signals/browserify.js";
import { detectBunBundler } from "./signals/bun.js";
import { detectEsbuild } from "./signals/esbuild.js";
import { detectMinifier } from "./signals/minifier.js";
import { detectParcel } from "./signals/parcel.js";
import { detectWebpack } from "./signals/webpack.js";
import type {
  BundlerType,
  BundlerDetectionResult,
  DetectionSignal,
  DetectionTier,
  MinifierType
} from "./types.js";

/** Maximum bytes to scan for signals (16KB) */
const SCAN_LIMIT = 16 * 1024;

/** Confidence ordering used for tier-aware signal selection (higher wins). */
const TIER_RANK: Record<DetectionTier, number> = {
  definitive: 2,
  likely: 1,
  unknown: 0
};

type BundlerDetector = (code: string) => DetectionSignal[];

const BUNDLER_DETECTORS: BundlerDetector[] = [
  detectWebpack,
  detectBrowserify,
  detectEsbuild,
  detectParcel,
  detectBunBundler
];

export function detectBundle(code: string): BundlerDetectionResult {
  const slice = code.slice(0, SCAN_LIMIT);

  const allSignals: DetectionSignal[] = [];

  // Run bundler detectors
  for (const detector of BUNDLER_DETECTORS) {
    allSignals.push(...detector(slice));
  }

  // Run minifier detectors
  allSignals.push(...detectMinifier(slice));

  // Pick bundler from definitive signals
  const definiteBundlerSignals = allSignals.filter(
    (s) => s.bundler && s.tier === "definitive"
  );

  let bundler: BundlerDetectionResult["bundler"];
  if (definiteBundlerSignals.length > 0) {
    // Use the first definitive bundler signal (they should agree)
    const type = definiteBundlerSignals[0].bundler as BundlerType;
    bundler = { type, tier: "definitive" };
  } else {
    bundler = { type: "unknown", tier: "unknown" };
  }

  // Pick minifier by highest tier, not array order: a distinctive esbuild/bun/swc
  // signal ("likely") outranks the generic terser fallback ("unknown"). Ties keep
  // the earliest signal in detector order, so selection is deterministic.
  const minifierSignals = allSignals.filter((s) => s.minifier);
  let minifier: BundlerDetectionResult["minifier"];
  if (minifierSignals.length > 0) {
    const best = minifierSignals.reduce((a, b) =>
      TIER_RANK[b.tier] > TIER_RANK[a.tier] ? b : a
    );
    minifier = { type: best.minifier as MinifierType, tier: best.tier };
  } else {
    minifier = { type: "unknown", tier: "unknown" };
  }

  return { bundler, minifier, signals: allSignals };
}

File: crates/humanify-core/src/detect.rs
//! Stage 1: bundle detection (TS: src/detection/detect.ts, WPB.1) — which
//! bundler wrapped the input and which minifier produced it, from string
//! signals in the first 16K UTF-16 code units. A pipeline stage, not a CLI
//! concern (02 §2): the CLI's `--bundler` / `--minifier` overrides are
//! applied on top of this verdict by the caller.
//!
//! Gate: `humanify detect <input>` prints `serde_json::to_string` of the
//! verdict, byte-compared with the TS `JSON.stringify(detectBundle(code))`
//! (test/parity/wpb1-detect-probe.ts).

pub mod js_text;
pub mod signals;

use humanify_model::detection::{
    BundlerDetectionResult, BundlerType, BundlerVerdict, DetectionSignal, DetectionTier,
    MinifierType, MinifierVerdict,
};

use self::js_text::js_prefix;
use self::signals::{
    detect_browserify, detect_bun_bundler, detect_esbuild, detect_minifier, detect_parcel,
    detect_webpack,
};

/// TS `SCAN_LIMIT`: 16K, in UTF-16 code units (`code.slice(0, SCAN_LIMIT)`).
const SCAN_LIMIT: usize = 16 * 1024;

/// TS `BUNDLER_DETECTORS`, in order (the order decides which definitive
/// signal wins when several fire).
const BUNDLER_DETECTORS: [fn(&str) -> Vec<DetectionSignal>; 5] = [
    detect_webpack,
    detect_browserify,
    detect_esbuild,
    detect_parcel,
    detect_bun_bundler,
];

/// TS `detectBundle`.
pub fn detect_bundle(code: &str) -> BundlerDetectionResult {
    let slice = js_prefix(code, SCAN_LIMIT);
    let mut signals: Vec<DetectionSignal> =
        BUNDLER_DETECTORS.iter().flat_map(|d| d(slice)).collect();
    signals.extend(detect_minifier(slice));
    BundlerDetectionResult {
        bundler: pick_bundler(&signals),
        minifier: pick_minifier(&signals),
        signals,
    }
}

/// The FIRST definitive bundler signal, else unknown.
fn pick_bundler(signals: &[DetectionSignal]) -> BundlerVerdict {
    let first = signals
        .iter()
        .find(|s| s.bundler.is_some() && s.tier == DetectionTier::Definitive)
        .and_then(|s| s.bundler);
    match first {
        Some(kind) => BundlerVerdict {
            kind,
            tier: DetectionTier::Definitive,
            version: None,
        },
        None => BundlerVerdict {
            kind: BundlerType::Unknown,
            tier: DetectionTier::Unknown,
            version: None,
        },
    }
}

/// The highest-tier minifier signal; ties keep the EARLIEST (the TS
/// `reduce` replaces the accumulator only on a strictly greater rank).
fn pick_minifier(signals: &[DetectionSignal]) -> MinifierVerdict {
    let best = signals
        .iter()
        .filter_map(|s| s.minifier.map(|m| (m, s.tier)))
        .reduce(|a, b| if b.1 > a.1 { b } else { a });
    match best {
        Some((kind, tier)) => MinifierVerdict { kind, tier },
        None => MinifierVerdict {
            kind: MinifierType::Unknown,
            tier: DetectionTier::Unknown,
        },
    }
}
